### PR TITLE
feat: mock_call with dynamic return data

### LINKED ISSUE #2861: `mock_call` with dynamic return data
### Which component is your feature related to?

Forge

### Feature Request

`mock_call` cheatcode accepts static `ret_data`. It could be useful if it was possible to define a function in in that can receive the calldata and dynamically return different ret_data.

### PR DESCRIPTION
**Pull request by ptisserand**
Closes #2861 

## Introduced changes

- Add `mock_call_when`, `start_mock_call_when` and `stop_mock_call_when` to support "dynamic" return value  when mocking a contract entry point.

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: snforge_std/src/cheatcodes.cairo
[the remark this answers is not in the document]
Comment (franciszekjob) [reply snforge_std/src/cheatcodes.cairo:1]: @cptartur I agree with this approach, we won't have breaking change + users won't be forced to pass `MockCallData::Any` frequently (which would happen in case of having current impl as `mock_call`).

[the remark this answers is not in the document]
Comment (ptisserand) [reply snforge_std/src/cheatcodes.cairo:1]: Clear, source code updated to remove the breaking change

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs [hunk 1]
```diff
@@ -268,11 +270,25 @@ fn get_mocked_function_cheat_status<'a>(
...
-        .and_then(|contract_functions| contract_functions.get_mut(&call.entry_point_selector))
+    {
+        None => None,
+        Some(contract_functions) => {
+            let call_data_hash = poseidon_hash_many(call.calldata.0.iter());
+            let key = (call.entry_point_selector, call_data_hash);
+            let key_zero = (call.entry_point_selector, Felt::zero());
+
+            match contract_functions.get(&key) {
+                Some(CheatStatus::Cheated(_, CheatSpan::TargetCalls(0))) => {
```
Comment (cptartur) [inline crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:284]: After calling `get_mocked_function_cheat_status`, we call `.decrement_cheat_span` on the returned value. It changes the CheatSpan to `Uncheated` after it decreases to `0`. 

I don't think there can be a case where `TargetCalls` is equal to 0 unless explicitly set to this value somewhere.

Comment (cptartur) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:284]: I assume this logic match arm was added to handle the case where entrypoint for specific calldata is no longer cheated but for any calldata still is, but it should already be handled by `_` case anyway.

Comment (cptartur) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:284]: Turns out technically it could be set to `0` but it shouldn't be allowed.

Created #2927

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/cheatcodes/mock_call.rs [hunk 2]
```diff
@@ -9,26 +11,30 @@ impl CheatnetState {
...
         ret_data: &[Felt],
         span: CheatSpan,
     ) {
         let contract_mocked_functions = self.mocked_functions.entry(contract_address).or_default();
-
-        contract_mocked_functions.insert(
-            function_selector,
-            CheatStatus::Cheated(ret_data.to_vec(), span),
-        );
+        let call_data_hash = match call_data {
```
Comment (cptartur) [inline crates/cheatnet/src/runtime_extensions/forge_runtime_extension/cheatcodes/mock_call.rs:19]: Nitpick, but in most places in the code we already call it `calldata`, not `call_data`.

File: crates/forge/tests/integration/mock_call.rs
Comment (cptartur) [inline crates/forge/tests/integration/mock_call.rs:1]: I think we should also have test cases where `mock_call_when` is mixed with `mock_call`, these are the ones most likely to break/have some weird edge cases.

File: docs/src/appendix/cheatcodes/mock_call_when.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+`MockCallData` is an enum used to specify for which call data the contract entry point will be mocked.
+- `Any` mock the contract entry point for any call data.
+- `Values` mock the contract entry point only for this call data.
+
+## `mock_call_when`
+> `fn mock_call_when<T, impl TSerde: serde::Serde<T>, impl TDestruct: Destruct<T>>(
+>   contract_address: ContractAddress, function_selector: felt252, call_data: MockCallData, ret_data: T, n_times: u32
+> )`
+
+Mocks contract call to a `function_selector` of a contract at the given address, with the given call data, for `n_times` first calls that are made
```
Comment (cptartur) [inline docs/src/appendix/cheatcodes/mock_call_when.md:23]: Here and in other places

```suggestion
Mocks contract call to a `function_selector` of a contract at the given address, with the given calldata, for `n_times` first calls that are made 
```

File: snforge_std/src/cheatcodes.cairo [hunk 4]
```diff
@@ -20,6 +20,36 @@ pub enum CheatSpan {
...
+    }
+
+    fn serialize(self: @MockCallData, ref output: Array<felt252>) {
+        match self {
+            MockCallData::Any => Option::<Span<felt252>>::None.serialize(ref output),
+            MockCallData::Values(data) => Option::Some(*data).serialize(ref output),
+        }
+    }
+}
+
```
Comment (cptartur) [inline snforge_std/src/cheatcodes.cairo:52]: Why not just derive it?

```suggestion
/// Enum used to specify the call data that should be matched when mocking a contract call.
#[derive(Copy, Drop, PartialEq, Clone, Debug, Serde)]
pub enum MockCallData {
    /// Matches any call data.
    Any,
    /// Matches the specified serialized call data.
    Values: Span<felt252>,
}
```

**Review by cptartur — CHANGES_REQUESTED** [review]

File: snforge_std/src/cheatcodes.cairo (hunk 4 above)
Comment (ptisserand) [reply snforge_std/src/cheatcodes.cairo]: Source code updated in cheatnet crate introduce `MockCalldata` enum

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/cheatcodes/mock_call.rs (hunk 2 above)
Comment (ptisserand) [reply crates/cheatnet/src/runtime_extensions/forge_runtime_extension/cheatcodes/mock_call.rs]: Source code updated to use `calldata` instead of `call data`

File: docs/src/appendix/cheatcodes/mock_call_when.md (hunk 3 above)
Comment (ptisserand) [reply docs/src/appendix/cheatcodes/mock_call_when.md]: Source code updated to use `calldata` instead of `call data`

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs (hunk 1 above)
Comment (ptisserand) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:284]: I'm not sure to understand what should I do here: remove my workaround?

Comment (cptartur) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:284]: You don't have to match for the case where cheat span is `CheatSpan::TargetCalls(0)`. Technically, user could have somehow created it like this (as explained in #2927), but during normal operation it will either be `CheatSpan::TargetCalls(some value > 0)` or `CheatStatus::Uncheated`.

Comment (cptartur) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:284]: See this https://github.com/foundry-rs/starknet-foundry/blob/f2e7d0d4077a654992914a688575a85bbe57b9a9/crates/cheatnet/src/state.rs#L141

File: crates/forge/tests/integration/mock_call.rs [hunk 5]
```diff
@@ -207,3 +207,376 @@ fn mock_calls() {
...
+        )
+        .unwrap()
+    );
+
+    let result = run_test_case(&test);
+    assert_passed(&result);
+}
+
+#[test]
+fn mock_calls_when_mixed() {
```
Comment (cptartur) [inline crates/forge/tests/integration/mock_call.rs:503]: Can we add some cases where the same tests uses both `mock_call` and `mock_call_when`?

E.g. in case where you do `start_mock_call_when(MockCalldata::Any, ...)` and then `stop_mock_call(...)` and similar?

**Comment by franciszekjob** [comment]
Hey @ptisserand 👋 , are you going to continue working on this PR?

**Comment by ptisserand** [comment]
> Hey @ptisserand 👋 , are you going to continue working on this PR?

Hi @franciszekjob, sorry I completely forgot about this PR when working on other projects :(
If it's not blocking on your side, I can work on it in the incoming week: what do you think ?

**Comment by franciszekjob** [comment]
@ptisserand sure thing, no hurry!

**Comment by franciszekjob** [comment]
Hi @ptisserand 👋 , kindly asking if there's any update on this?

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs (hunk 1 above)
Comment (ptisserand) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs]: Workaround removed

**Comment by cptartur** [comment]
Hey @ptisserand is this ready for re-review?

**Comment by ptisserand** [comment]
> Hey @ptisserand is this ready for re-review?

Not yet, I have only managed conflicts with master branch.
I didn't yet add requested additional test cases.

**Comment by franciszekjob** [comment]
Hi @ptisserand 👋 , any update on this?

**Comment by ptisserand** [comment]
Hi @franciszekjob, I have added some tests cases.
I think this PR could be reviewed again.

**Comment by ptisserand** [comment]
Hi @cptartur @franciszekjob, could you please review this PR ?

File: crates/forge/tests/integration/mock_call.rs (hunk 5 above)
Comment (franciszekjob) [reply crates/forge/tests/integration/mock_call.rs:504]: How about this @ptisserand ?

File: docs/src/appendix/cheatcodes/mock_call_when.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+>   contract_address: ContractAddress, function_selector: felt252, calldata: MockCalldata, ret_data: T
+> )`
+
+Mocks contract call to a `function_selector` of a contract at the given address, with the given calldata, indefinitely.
+See `mock_call_when` for comprehensive definition of how it can be used.
+
+
+### `stop_mock_call_when`
+
+> `fn stop_mock_call_when(contract_address: ContractAddress, function_selector: felt252, calldata: MockCalldata)`
```
Comment (franciszekjob) [inline docs/src/appendix/cheatcodes/mock_call_when.md:40]:
```suggestion
## `stop_mock_call_when`
> `fn stop_mock_call_when(contract_address: ContractAddress, function_selector: felt252, calldata: MockCalldata)`
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -362,6 +362,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Rust is no longer required to use `snforge` if using Scarb >= 2.10.0 on supported platforms - precompiled `snforge_scarb_plugin` plugin binaries are now published to [package registry](https://scarbs.xyz) for new versions.
 - Added a suggestion for using the `--max-n-steps` flag when the Cairo VM returns the error: `Could not reach the end of the program. RunResources has no remaining steps`.
+- `mock_call_when`, `start_mock_call_when`, `stop_mock_call_when` cheatcodes.
```
Comment (franciszekjob) [inline CHANGELOG.md]: It should be under unreleased version.

File: docs/src/appendix/cheatcodes/mock_call_when.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
+# `mock_call_when`
+
+Cheatcodes mocking contract entry point calls:
```
Comment (franciszekjob) [inline docs/src/appendix/cheatcodes/mock_call_when.md]: It's the same as description of `mock_call`. Maybe let's emphasize the difference?

File: docs/src/appendix/cheatcodes/mock_call_when.md
Comment (franciszekjob) [inline docs/src/appendix/cheatcodes/mock_call_when.md:1]: It would be really nice to add an example of usage, similarly how it's done [here](https://foundry-rs.github.io/starknet-foundry/appendix/cheatcodes/mock_call.html#example) 😄 
You can create and issue for that, link it here and if you have time open - a separate PR.

File: CHANGELOG.md (hunk 7 above)
Comment (ptisserand) [reply CHANGELOG.md]: Done

File: docs/src/appendix/cheatcodes/mock_call_when.md (hunk 6 above)
Comment (ptisserand) [reply docs/src/appendix/cheatcodes/mock_call_when.md:40]: Done

File: docs/src/appendix/cheatcodes/mock_call_when.md (hunk 8 above)
Comment (ptisserand) [reply docs/src/appendix/cheatcodes/mock_call_when.md]: Updated with `Cheatcodes mocking contract entry point calls based on calldata:`
